### PR TITLE
Replace the v1 api variant search with BigQuery call

### PR DESCRIPTION
**Pull request by anamanolache**

### CODE REVIEW

File: app.yaml [hunk 1]
```diff
@@ -12,8 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-application: beacon-go
-version: 1
+#application: beacon-go
```
Comment (kemp-google) [inline app.yaml]: just remove, not comment out

File: beacon.go [hunk 2]
```diff
@@ -19,108 +19,95 @@ package beacon
...
 
 type beaconConfig struct {
-	variantSetIds []string
+	projectID string
+	table     string
 }
 
 var config = beaconConfig{
-	variantSetIds: []string{"3049512673186936334"},
+	projectID: "project-id",
```
Comment (kemp-google) [inline beacon.go]: this should be read from the YAML as environment variables, eg, in app.yaml:

env_variables:
  TABLE: genomics-public-data.platinum_genomes.variants

It's also possible that GOOGLE_CLOUD_PROJECT is already set by the runtime with the project ID

File: beacon.go [hunk 3]
```diff
@@ -19,108 +19,95 @@ package beacon
...
 
-func requestToSearch(r *http.Request) (*genomics.SearchVariantsRequest, string, error) {
-	allele := r.FormValue("allele")
-	search := &genomics.SearchVariantsRequest{
-		ReferenceName: r.FormValue("chromosome"),
-		VariantSetIds: config.variantSetIds,
+func handler(w http.ResponseWriter, r *http.Request) {
+	refName, allele, coord, err := parseInput(r)
+	if err != nil {
+		http.Error(w, "Failed parsing parameters", http.StatusBadRequest)
```
Comment (kemp-google) [inline beacon.go]: You need to return after writing a response with http.Error

It's better to do more like the original code where you have a function that returns an error and that calls http.Error as appropriate

File: app.yaml (hunk 1 above)
Comment (anamanolache) [reply app.yaml]: done

File: beacon.go (hunk 3 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 2 above)
Comment (anamanolache) [reply beacon.go]: done. I didn't find GOOGLE_CLOUD_PROJECT preset so I set it

**Comment by anamanolache** [comment]
I missed that @arostamianfar, I'll update it to match the specs. Thanks

**Comment by kemp-google** [comment]
@arostamianfar we can make updates but it will go into a separate PR

File: app.yaml [hunk 4]
```diff
@@ -12,11 +12,13 @@
...
-version: 1
 runtime: go
 api_version: go1
 
 handlers:
 - url: /.*
   script: _go_app
+
+env_variables:
+  GOOGLE_CLOUD_PROJECT: project-id
```
Comment (kemp-google) [inline app.yaml]: I'd leave this blank or change it to something like 'change-to-your-project-id'

Otherwise errors will be like:

"can't find table project-id/foo/bar" and it is less clear

Comment (kemp-google) [reply app.yaml]: Even better would be to *not* set these by default [just comment them out] and throw an error or something on the request

**Comment by arostamianfar** [comment]
A separate PR SGTM!

File: app.yaml [hunk 5]
```diff
@@ -12,11 +12,13 @@
...
+  TABLE: genomics-public-data.platinum_genomes.variants
```
Comment (pgrosu) [inline app.yaml]: If the `GOOGLE_CLOUD_PROJECT` is blank then the same logic should go for the TABLE.  Actually `TABLE` is quite vague and it should infer from the project the Beacon tables available to that project.

File: beacon.go [hunk 6]
```diff
@@ -17,110 +17,118 @@
...
-		return nil, "", errors.New("missing reference name")
+func handler(w http.ResponseWriter, r *http.Request) {
+	if config.projectID == "" {
+		http.Error(w, "Mandatory variable GOOGLE_CLOUD_PROJECT not set", http.StatusInternalServerError)
+		return
 	}
-	if allele == "" {
-		return nil, "", errors.New("missing allele")
+	if config.table == "" {
+		http.Error(w, "Mandatory variable TABLE not set", http.StatusInternalServerError)
```
Comment (pgrosu) [inline beacon.go]: What is a variable table?  It is a variant table.

Comment (pgrosu) [reply beacon.go]: I see what you guys are doing but it might be confusing to a Bioinformatician.

**Comment by pgrosu** [comment]
@anamanolache @arostamianfar @kemp-google If you are trying to maintain conformance, please try to use the releases:

https://github.com/ga4gh-beacon/specification/releases

Thanks,
Paul

File: app.yaml [hunk 7]
```diff
@@ -12,11 +12,13 @@
...
 runtime: go
 api_version: go1
 
 handlers:
 - url: /.*
   script: _go_app
+
+env_variables:
+#  GOOGLE_CLOUD_PROJECT: change-to-your-project-id
+#  TABLE: genomics-public-data.platinum_genomes.variants
```
Comment (kemp-google) [inline app.yaml]: GOOGLE_BIGQUERY_TABLE ?

File: beacon.go [hunk 8]
```diff
@@ -17,110 +17,185 @@
...
+
+func newInvalidInputError(context string, err error) error {
+	return newApiError("InvalidInput", http.StatusBadRequest, context, err)
+}
+
+func newDataAccessError(context string, err error) error {
+	return newApiError("DataAccessError", http.StatusInternalServerError, context, err)
+}
+
+// writeError writes either a JSON object or bare HTTP error describing err to
```
Comment (kemp-google) [inline beacon.go]: The JSON object encoding an error is specific to htsget and doesn't apply here.

It looks like Beacon does define a BeaconError that is similar, and we could write one, but that is outside the scope of this change.

File: beacon.go [hunk 9]
```diff
@@ -17,110 +17,185 @@
...
+	}
+	return result.Count > 0, nil
+}
+
+func validateServerConfig() error {
+	if config.projectID == "" {
+		return newInvalidConfigError("validating GOOGLE_CLOUD_PROJECT value", errors.New("value is mandatory"))
+	}
+	if config.table == "" {
+		return newInvalidConfigError("validating TABLE value", errors.New("value is mandatory"))
```
Comment (kemp-google) [inline beacon.go]: Something simpler like

fmt.Errorf("%s must be specified", tableConfigurationKey)

will be sufficient since it will result in a generic server error.  We only need to be specific about error codes where it makes sense and is defined by the specification.

(Applies to the project too)

File: beacon.go (hunk 8 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 9 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go [hunk 10]
```diff
@@ -17,110 +17,174 @@
...
+	return result.Count > 0, nil
 }
 
-func handler(w http.ResponseWriter, r *http.Request) {
-	c := appengine.NewContext(r)
-	search, allele, err := requestToSearch(r)
-	if err != nil {
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+func validateServerConfig() error {
```
Comment (kemp-google) [inline beacon.go]: You should just return a bare error from this function

And again, you don't need to use any special error functions here: a server error will be the default kind of error.  You only need to use a specific error type when you want a non-500 error code.

File: beacon.go [hunk 11]
```diff
@@ -17,110 +17,174 @@
...
+		LIMIT 1`,
+		fmt.Sprintf("`%s`", config.tableId),
+		refName,
+		coord,
+		coord+1,
+		allele)
+
+	bqclient, err := bigquery.NewClient(ctx, config.projectID)
+	if err != nil {
+		return false, newDataAccessError("Creating a BigQuery client", err)
```
Comment (kemp-google) [inline beacon.go]: Error strings are lowercase (eg "creating bigquery client")

File: beacon.go [hunk 12]
```diff
@@ -17,110 +17,174 @@
...
+	it, err := bqclient.Query(query).Read(ctx)
+	if err != nil {
+		return false, newDataAccessError("Querying database", err)
 	}
 
-	return search, allele, nil
+	type Result struct {
+		Count int
+	}
+	var result Result
```
Comment (kemp-google) [inline beacon.go]: since you don't need the result type other places, you can just say

var result struct {
  Count int
}
...

File: app.yaml (hunk 7 above)
Comment (pgrosu) [reply app.yaml]: @anamanolache @kemp-google That's a bit better, thanks.

File: beacon.go [hunk 13]
```diff
@@ -17,110 +17,174 @@
...
 	enc.Indent("", "  ")
-	if err = enc.Encode(resp); err != nil {
-		http.Error(w, "Failed writing response", http.StatusInternalServerError)
+	if err := enc.Encode(resp); err != nil {
+		return newApiError(http.StatusInternalServerError, "Serializing response", err)
+	}
+	return nil
+}
+
+// apiError is used to capture errors that have been defined in the API.
```
Comment (kemp-google) [inline beacon.go]: This isn't accurate for this package.  You just need an httpError type

type httpError struct {
...
}

File: beacon.go [hunk 14]
```diff
@@ -17,110 +17,174 @@
...
+		coord+1,
+		allele)
+
+	bqclient, err := bigquery.NewClient(ctx, config.projectID)
+	if err != nil {
+		return false, newDataAccessError("Creating a BigQuery client", err)
+	}
+	it, err := bqclient.Query(query).Read(ctx)
+	if err != nil {
+		return false, newDataAccessError("Querying database", err)
```
Comment (pgrosu) [inline beacon.go]: probably here to too for lowercase

File: beacon.go [hunk 15]
```diff
@@ -17,110 +17,174 @@
...
-	}
+	resp.Exists = exists
 
 	w.Header().Set("Content-Type", "application/xml")
 	enc := xml.NewEncoder(w)
 	enc.Indent("", "  ")
-	if err = enc.Encode(resp); err != nil {
-		http.Error(w, "Failed writing response", http.StatusInternalServerError)
+	if err := enc.Encode(resp); err != nil {
+		return newApiError(http.StatusInternalServerError, "Serializing response", err)
```
Comment (pgrosu) [inline beacon.go]: lowercase again

File: beacon.go (hunk 11 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 12 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 13 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 10 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go [hunk 16]
```diff
@@ -17,110 +17,164 @@
...
 }
 
-func requestToSearch(r *http.Request) (*genomics.SearchVariantsRequest, string, error) {
-	allele := r.FormValue("allele")
-	search := &genomics.SearchVariantsRequest{
-		ReferenceName: r.FormValue("chromosome"),
-		VariantSetIds: config.variantSetIds,
+func handler(w http.ResponseWriter, r *http.Request) {
+	if err := validateServerConfig(); err != nil {
+		writeError(w, fmt.Errorf("validating server config: %v", err))
```
Comment (kemp-google) [inline beacon.go]: s/config/configuration

Errors are for users and should be free of slang or shortforms

File: beacon.go [hunk 17]
```diff
@@ -17,110 +17,164 @@
...
 func init() {
 	http.HandleFunc("/", handler)
 }
 
-func requestToSearch(r *http.Request) (*genomics.SearchVariantsRequest, string, error) {
-	allele := r.FormValue("allele")
-	search := &genomics.SearchVariantsRequest{
-		ReferenceName: r.FormValue("chromosome"),
-		VariantSetIds: config.variantSetIds,
+func handler(w http.ResponseWriter, r *http.Request) {
```
Comment (kemp-google) [inline beacon.go]: *We can ignore this for this change, and do it in a follow up*

It's nice to keep the HTTP machinery away from the logic.  See how it's done with the api package in the htsget server.  This allows you to have regular functions (that take regular arguments and return regular values and errors).

The original code here was closer to this ideal.  Let's get this in and then we can go through how to refactor.

File: beacon.go [hunk 18]
```diff
@@ -17,110 +17,164 @@
...
+	return nil
+}
 
-	client, err := google.DefaultClient(c, genomics.GenomicsReadonlyScope)
-	if err != nil {
-		http.Error(w, "Invalid server configuration", http.StatusInternalServerError)
+func parseInput(r *http.Request) (string, string, int64, error) {
+	refName := r.FormValue("chromosome")
+	if refName == "" {
+		return "", "", 0, newBadRequestError("parsing chromosome name", errors.New("value is required"))
```
Comment (kemp-google) [inline beacon.go]: Here, again, you should return bare errors and let the calling function annotate it with context:

{
  ...
   return "", "", 0, errors.New("missing chromosome name")
  ...
}

if err := parseInput(...); err != nil {
   return newBadRequestError("parsing input parameters", err)
}

File: beacon.go [hunk 19]
```diff
@@ -17,110 +17,164 @@
...
+	return &apiError{code, fmt.Errorf("%s: %v", context, err)}
+}
+
+func newBadRequestError(context string, err error) error {
+	return newHttpError(http.StatusBadRequest, context, fmt.Errorf("invalid input: %v", err))
+}
+
+// writeError writes a bare HTTP error describing err to w.
+func writeError(w http.ResponseWriter, err error) {
+	if err, ok := err.(*apiError); ok {
```
Comment (kemp-google) [inline beacon.go]: code := http.StatusInternalServerError
if err, ok := err.(*apiError); ok {
  code = err.code
}
http.Error(w, fmt.Sprintf("%s: %v", http.StatusText(code), err), code)

Then you can drop `writeHTTPError`

File: beacon.go (hunk 17 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 16 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 18 above)
Comment (anamanolache) [reply beacon.go]: done

File: beacon.go (hunk 19 above)
Comment (anamanolache) [reply beacon.go]: not applicable since I reverted the errors to what they used to be

File: beacon.go [hunk 20]
```diff
@@ -17,110 +17,134 @@
...
 	enc := xml.NewEncoder(w)
 	enc.Indent("", "  ")
-	if err = enc.Encode(resp); err != nil {
-		http.Error(w, "Failed writing response", http.StatusInternalServerError)
+	if err := enc.Encode(resp); err != nil {
+		return fmt.Errorf("serializing response: %v", err)
 	}
-}
+	return nil
+}
```
Comment (kemp-google) [inline beacon.go]: Did you run this through gofmt?  I thought it forced a newline at the end of the file.
[In either case, add a newline]

Comment (anamanolache) [reply beacon.go]: forgot to. done

File: README.md [hunk 21]
```diff
@@ -20,9 +20,9 @@ In order to setup and deploy this application, you will need:
 - Clone this repo.
   - `git clone git@github.com:googlegenomics/beacon-go.git`
 - Edit the configuration.
-  - In `beacon.go`, edit the value of `variantSetIds` to reference your data.
+  - In `app.yaml`, edit the values of `GOOGLE_CLOUD_PROJECT` and `GOOGLE_BIGQUERY_TABLE` to reference your data.
 - Deploy.
-  - `goapp deploy -application <project name> app.yaml`
+  - `gcloud app deploy app.yaml --project <project name>`
```
Comment (anamanolache) [inline README.md]: @kemp-google, can you please review this line before I merge, it has been changed in a parallel MR and I had to resolve the conflict
